src/include/sof/uart.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2017-2018, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __INCLUDE_UART_H__
+#define __INCLUDE_UART_H__
+
+#include <errno.h>
+#include <stdint.h>
+
+struct uart_platform_data {
+	uint32_t port;		/* register base address */
+	unsigned int baud;	/* baud rate */
+	int irq;		/* interrupt number if >= 0 */
+	const char *irq_dev;	/* name of the interrupt controller */
+};
+
+struct uart {
+	const struct uart_platform_data *pdata;
+	const struct uart_ops *ops;
+};
+
+struct uart_ops {
+	void (*write_word)(struct uart *dev, uint32_t word);
+	int (*write_nowait)(struct uart *dev, const char *data, uint32_t size);
+	int (*write)(struct uart *dev, const char *data, uint32_t size);
+};
+
+static inline void uart_write_word(struct uart *uart, uint32_t word)
+{
+	if (uart)
+		uart->ops->write_word(uart, word);
+}
+
+static inline int uart_write(struct uart *uart,
+			     const char *data, uint32_t size)
+{
+	return uart ? uart->ops->write(uart, data, size) : -ENODEV;
```

Comment:
Can you also change this to check for write e.g.
```
if (uart && uart->ops->write)
        uart->ops->write();
```
As some drivers may not implement all the ops. Ditto for others.